lmdeploy/pytorch/engine/model_agent.py

```diff
@@ -566,6 +566,32 @@ def raise_error(self, default_error: Exception):
                 raise err
 
 
+def _get_model_memory_usage(model: torch.nn.Module) -> int:
+    """get model memory usage."""
+    size = 0
+    for _, param in model.named_parameters():
+        size += param.element_size() * param.numel()
+    for _, buf in model.named_buffers():
+        size += buf.element_size() * param.numel()
+    return size
+
+
+def _create_device_map(model: torch.nn.Module,
+                       world_size: int,
+                       device_map: dict = None):
+    """Distribute params to each devices."""
+    if device_map is None:
+        device_map = dict()
+    device_id = 0
+    for name, _ in model.named_parameters():
+        device_map[name] = device_id
+        device_id = (device_id + 1) % world_size
+    for name, _ in model.named_buffers():
```

Comment:
Will this method make some GPUs heavily loaded and others are not?

Reply:
In some special case, yes.
For example, if `tp==2` and weight size is `[large, small, large, small, ...]`, device0 would load all large weights and device1 would load all small weights. But I think this is almost not possible for LLM.